backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/s3/S3Service.java

```diff
@@ -60,257 +47,45 @@
-    // using the sessionCredentials object.
-    s3Client =
-        AmazonS3ClientBuilder.standard()
-            .withCredentials(new AWSStaticCredentialsProvider(awsCredentials))
-            .withRegion(clientRegion)
-            .build();
-    try {
-      s3Client.doesBucketExistV2(bucketName);
-      LOGGER.debug("s3client refreshed with temporary credentials");
-      return true;
-    } catch (AmazonServiceException e) {
-      ModelDBUtils.logAmazonServiceExceptionErrorCodes(LOGGER, e);
-    } catch (SdkClientException e) {
-      LOGGER.warn(e.getMessage());
-    }
-    return false;
-  }
-
-  private void RefreshCredentialsAndSchedule(String region, int retryTimeForAwsCredentials) {
-    LOGGER.debug("fetching token for s3 access");
-    if (task != null) {
-      task.cancel();
-      task = null;
-    }
-
-    LOGGER.debug(
-        "credentials before refresh {}",
-        temporarySessionCredentials != null ? temporarySessionCredentials.hashCode() : null);
-
-    preScheduleTimer(region, retryTimeForAwsCredentials);
-  }
-
-  public static synchronized void preScheduleTimer(String region, int retryTimeForAwsCredentials) {
-    if (task == null) {
-      task = new FetchTemporaryS3Token(region, retryTimeForAwsCredentials);
-      task.run();
-    }
-
-    if (temporarySessionCredentials == null) {
-      LOGGER.warn(
-          "After retrying to fetch new credentials till {} second, temporarySessionCredentials is not initialize",
-          retryTimeForAwsCredentials);
-      throw StatusProto.toStatusRuntimeException(
-          Status.newBuilder()
-              .setCode(Code.UNAVAILABLE_VALUE)
-              .setMessage("Getting problem to connect amazon service")
-              .build());
-    }
-  }
-
-  public static synchronized void scheduleTimer(String region) {
-    task.cancel();
-
-    LOGGER.debug(
-        "credentials after refresh {}",
-        temporarySessionCredentials != null ? temporarySessionCredentials.hashCode() : null);
-    LOGGER.debug("fetched token for s3 access");
-
-    Date expiration = temporarySessionCredentials.getExpiration();
-    Date now = new Date();
-    long diffInSec = Math.abs(expiration.getTime() - now.getTime()) / 1000;
-    long delay = diffInSec / 2;
-    LOGGER.info("scheduled cron for credential refresh in {} seconds", delay);
-    task = new FetchTemporaryS3Token(region, Long.valueOf(delay).intValue());
-    ModelDBUtils.scheduleTask(task, delay, delay, TimeUnit.SECONDS);
-    LOGGER.debug("scheduled periodic task to fetch token for s3 access");
+    this.bucketName = cloudBucketName;
   }
 
-  private Boolean doesBucketExist(String bucketName) {
-    try {
-      return s3Client.doesBucketExistV2(bucketName);
+  private Boolean doesBucketExist(String bucketName) throws ModelDBException {
+    try (RefCountedS3Client client = s3Client.getRefCountedClient()) {
+      return client.getClient().doesBucketExistV2(bucketName);
     } catch (AmazonServiceException e) {
```

Comment:
When the client/user calls the MDB backend for artifact store related endpoints and at that time s3Client is not valid then should we call refresh credentials and s3Client or throw an exception message to the client/user? we are doing it before your changes to protect that client call from the exception on the catch block https://github.com/VertaAI/modeldb/pull/1584/files#diff-4e6e61b32315bc91cd94ab22fc503186db0883222f37ac84e92a7910efabe7cfL264-L270

Reply:
I'd suggest exception. It should _never_ happen as we try to refresh like crazy halfway from the session life already. If we weren't able to refresh, then it's probably something wrong with aws and trying again won't help much.

Reply:
Ok, it makes sense.